### PR TITLE
Update wire protocol to support sending both long and short offers

### PR DESCRIPTION
**Pull request by klochowicz**
This is a refactor, which aimed to preserve the current functionality - the only
difference from API perspective is the nature of choosing whether the maker
wants to go short or long (now it's done by setting `price_long` if maker
wants to go long, before it was choosing `position_maker`).

### CODE REVIEW

File: daemon/src/maker_cfd.rs [hunk 1]
```diff
@@ -75,25 +83,76 @@ pub struct RejectRollover {
...
-#[derive(Clone, Copy)]
-pub struct TakerConnected {
-    pub id: Identity,
+impl OfferParams {
+    fn pick_oracle_event_id(settlement_interval: Duration) -> BitMexPriceEventId {
+        oracle::next_announcement_after(time::OffsetDateTime::now_utc() + settlement_interval)
+    }
+
+    pub fn create_long_order(&self, settlement_interval: Duration) -> Option<Order> {
+        self.price_long.map(|price_long| {
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd.rs:113]: You can `?` `self.price_long` here then you can spare yourself the `map`.

File: model/src/cfd.rs [hunk 2]
```diff
@@ -109,6 +109,62 @@ impl From<Uuid> for OrderId {
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let long_order_id = self.long.as_ref().map(|o| o.id);
+        let short_order_id = self.long.as_ref().map(|o| o.id);
+        let tx_fee_rate = self.tx_fee_rate;
+        let funding_rate = self.funding_rate;
+
+        write!(
+            f,
+            "short order_id={short_order_id:?}, long order_id={long_order_id:?}, tx_fee_rate={tx_fee_rate:?}, funding_rate={funding_rate}"
+        )
```
Comment (thomaseizinger) [inline model/src/cfd.rs]: Why not use the usual structure of `Debug` here? You can create it with the helper functions on `Formatter`.

File: daemon/src/maker_cfd.rs [hunk 3]
```diff
@@ -75,25 +83,76 @@ pub struct RejectRollover {
...
 
+#[derive(Clone, Copy)]
+pub struct TakerConnected {
+    pub id: Identity,
+}
+
+#[derive(Clone, Copy)]
+pub struct TakerDisconnected {
+    pub id: Identity,
+}
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd.rs:94]: Nit: The diff would be a bit more minimal if you wouldn't move these!

File: model/src/cfd.rs [hunk 4]
```diff
@@ -109,6 +109,62 @@ impl From<Uuid> for OrderId {
...
+    }
+}
+
+impl MakerOffers {
+    pub fn take_offer_by_order_id(&self, id: OrderId) -> Option<Order> {
+        if let Some(long) = self.long {
+            if long.id == id {
+                self.long.expect("be still valid").replicate();
```
Comment (thomaseizinger) [inline model/src/cfd.rs]: You are dropping the return value of `replicate` here, yes? That seems like a mistake :)

Comment (thomaseizinger) [reply model/src/cfd.rs]: I think `take_offer_by_order_id` should not become `&mut self`. We are doing the replicate so we shouldn't need to do it in here.

File: daemon/src/maker_cfd.rs [hunk 5]
```diff
@@ -261,42 +320,43 @@ where
...
-            .send_async_safe(maker_inc_connections::BroadcastOrder(self.current_order))
+            .send_async_safe(maker_inc_connections::BroadcastOffers(self.current_offers))
             .await?;
 
-        #[allow(unused_qualifications)]
-        // Need to fully qualify `Option` because we have more than one `Update` message that
-        // contains an `Option<T>`.
-        self.projection
-            .send(projection::Update(self.current_order))
-            .await?;
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd.rs:299]: Lol, I appear to not even read / comprehend my own comments.

**Review by thomaseizinger — COMMENTED** [review]
Did a first pass, will have a closer look tomorrow morning!

File: daemon/src/maker_cfd.rs [hunk 6]
```diff
@@ -75,25 +83,76 @@ pub struct RejectRollover {
...
+                self.opening_fee,
+            )
+        })
+    }
 }
 
-#[derive(Clone, Copy)]
-pub struct TakerDisconnected {
-    pub id: Identity,
+// If I define an extension trait, can I still send the vanilla one?
```
Comment (da-kami) [inline daemon/src/maker_cfd.rs]: I think this comment can be removed

**Review by da-kami — APPROVED** [review]
nice! looking forward to plugging it into projection and the UI

File: model/src/cfd.rs (hunk 4 above)
Comment (klochowicz) [reply model/src/cfd.rs]: > You are dropping the return value of `replicate` here, yes? That seems like a mistake :)

you're right! it used to be `&mut self` for a bit, but this was really hard to comprehend and didn't work well! I forgot to remove it here, removing it now :)

File: daemon/src/maker_cfd.rs (hunk 3 above)
Comment (klochowicz) [reply daemon/src/maker_cfd.rs:94]: true, but then the impl block & factory function would be separated. I assumed I picked the lesser evil ;-) I will keep in mind minimising the diffs though for next time.

File: model/src/cfd.rs (hunk 2 above)
Comment (klochowicz) [reply model/src/cfd.rs]: > Why not use the usual structure of `Debug` here? You can create it with the helper functions on `Formatter`.

I'm afraid I don't know about the usual structure of `Debug` - would you mind throwing a link / explaining in more detail? I've only done it to keep our trace log that used to rely on order id working.

File: daemon/src/maker_cfd.rs (hunk 1 above)
Comment (klochowicz) [reply daemon/src/maker_cfd.rs:113]: `map` was clippy's suggestion - I made a "manual map implementation" in the first place 😅 

I guess I could, but this way I've got only one expression, and it's more clear IMHO - happy to change it if the other way is more often used around here! :)

**Review by rishflab — APPROVED** [review]
Sweet, lgtm. We need to also update the SSE feed to send both orders to the maker and taker. We can add that in a second PR.

File: model/src/cfd.rs (hunk 2 above)
Comment (thomaseizinger) [reply model/src/cfd.rs]: This function is your friend: https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct :)

File: daemon/src/maker_cfd.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/maker_cfd.rs:113]: I was suggesting to do this:

```rust
let price_long = self.price_long?;

Some(Order::new(
	  Position::Long,
	  price_long,
	  self.min_quantity,
	  self.max_quantity,
	  Origin::Ours,
	  Self::pick_oracle_event_id(settlement_interval),
	  settlement_interval,
	  self.tx_fee_rate,
	  self.funding_rate,
	  self.opening_fee,
))
```

Comment (thomaseizinger) [reply daemon/src/maker_cfd.rs:113]: `?` on an `Option` will bubble up the `None` variant.

Comment (klochowicz) [reply daemon/src/maker_cfd.rs:113]: Yes, I know - and I said that I prefer sparing the unnecessary assignment, and having a single expression if possible.

File: model/src/cfd.rs (hunk 2 above)
Comment (klochowicz) [reply model/src/cfd.rs]: wow, this is cool - thanks! :)

File: daemon/src/maker_cfd.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/maker_cfd.rs:113]: Okay :)

Using `?` is simpler for my head but I don't feel strongly about it.

File: daemon-tests/src/lib.rs [hunk 7]
```diff
@@ -423,16 +423,22 @@ pub fn dummy_quote() -> Quote {
     }
 }
 
+/// Offer params allowing a single position, either short or long
```
Comment (thomaseizinger) [inline daemon-tests/src/lib.rs]: Nit: triple-slash is rustdoc and that one has a convention that the first line is a summary description of the documentation: https://brson.github.io/2012/04/14/how-to-rustdoc#the-rules-of-rustdoc

A comment like this I would expect to be double-slash.

File: maker/src/routes.rs [hunk 8]
```diff
@@ -94,10 +94,11 @@ pub async fn maker_feed(
     }
 }
 
-/// The maker POSTs this to create a new CfdOrder
+/// The maker POSTs this to create new Offer params
```
Comment (thomaseizinger) [inline maker/src/routes.rs]:
```suggestion
/// The maker POSTs this to set the offer params
```

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**